include/deal.II/lac/affine_constraints.templates.h

```diff
@@ -191,6 +192,328 @@ AffineConstraints<number>::is_consistent_in_parallel(
 
 
 
+template <typename number>
+void
+AffineConstraints<number>::make_consistent_in_parallel(
+  const IndexSet &locally_owned_dofs,
+  const IndexSet &locally_relevant_dofs,
+  const MPI_Comm &mpi_communicator)
+{
+  if (Utilities::MPI::job_supports_mpi() == false ||
+      Utilities::MPI::n_mpi_processes(mpi_communicator) == 1)
+    return; // nothing to do, since serial
+
+#ifndef DEAL_II_WITH_MPI
+  AssertThrow(false, ExcNotImplemented()); // one should not come here
+  (void)locally_owned_dofs;
+  (void)locally_relevant_dofs;
+#else
+  Assert(sorted == false, ExcMatrixIsClosed());
+
+  using ConstraintType =
+    std::tuple<types::global_dof_index,
+               number,
+               std::vector<std::pair<types::global_dof_index, number>>>;
+
+  const unsigned int my_rank =
+    Utilities::MPI::this_mpi_process(mpi_communicator);
+
+  const auto compute_locally_relevant_constraints =
+    [&]() -> std::vector<ConstraintType> {
+    // The result vector filled step by step.
+    std::vector<ConstraintType> locally_relevant_constrains;
+
+    // helper function
+    const auto sort_constraints = [&]() {
+      std::sort(locally_relevant_constrains.begin(),
+                locally_relevant_constrains.end(),
+                [](const auto &a, const auto &b) {
+                  return std::get<0>(a) < std::get<0>(b);
+                });
+
+      locally_relevant_constrains.erase(
+        std::unique(locally_relevant_constrains.begin(),
+                    locally_relevant_constrains.end(),
+                    [](const auto &a, const auto &b) {
+                      return std::get<0>(a) == std::get<0>(b);
+                    }),
+        locally_relevant_constrains.end());
+    };
+
+    // 0) collect constrained indices of the current object
+    IndexSet constrained_indices(locally_owned_dofs.size());
+    for (const auto &line : this->get_lines())
+      constrained_indices.add_index(line.index);
```

Comment:
Did you check that this does not trigger a quadratic complexity part in `IndexSet`?